File: generated_code/GeminiPro/Python/086_SplayTree.py
class SplayTree:
    """
    A complete, self-contained implementation of a Splay Tree.

    This class implements a dictionary-like set for integer keys. Splay trees
    are self-balancing binary search trees with the property that recently
    accessed elements are quick to access again. Operations like insert, delete,
    and search have an amortized time complexity of O(log n).

    The core of this tree is the 'splay' operation, which moves a specific
    node to the root of the tree through a series of rotations. This is
    performed on every search, insert, and delete operation.
    """

    class _Node:
        """A private class representing a node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _right_rotate(self, x):
        """Performs a right rotation on the subtree rooted at x."""
        y = x.left
        x.left = y.right
        y.right = x
        return y

    def _left_rotate(self, x):
        """Performs a left rotation on the subtree rooted at x."""
        y = x.right
        x.right = y.left
        y.left = x
        return y

    def _splay(self, key):
        """
        Performs the top-down splay operation.

        This method brings the node with the given key to the root. If the key
        is not in the tree, the last accessed node (the would-be parent) is
        brought to the root. This method modifies `self.root` directly.
        """
        if not self.root:
            return

        # Use a dummy node to simplify linking of left and right subtrees
        dummy = self._Node(None)
        left_tree = dummy
        right_tree = dummy
        current = self.root

        while True:
            if key < current.key:
                if not current.left:
                    break
                # Zig-Zig case (right rotation)
                if key < current.left.key:
                    current = self._right_rotate(current)
                    if not current.left:
                        break
                # Link current node to the right tree
                right_tree.left = current
                right_tree = current
                current = current.left
            elif key > current.key:
                if not current.right:
                    break
                # Zig-Zig case (left rotation)
                if key > current.right.key:
                    current = self._left_rotate(current)
                    if not current.right:
                        break
                # Link current node to the left tree
                left_tree.right = current
                left_tree = current
                current = current.right
            else:  # Key is found
                break

        # Reassemble the tree
        left_tree.right = current.left
        right_tree.left = current.right
        current.left = dummy.right
        current.right = dummy.left
        
        self.root = current

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the node is splayed to the root. If the key
        is new, it is inserted and becomes the new root.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        # Splay the tree to bring the closest node to the root
        self._splay(key)

        # If key is already at the root, insertion is complete
        if self.root.key == key:
            return

        new_node = self._Node(key)
        if key < self.root.key:
            new_node.right = self.root
            new_node.left = self.root.left
            self.root.left = None
        else:  # key > self.root.key
            new_node.left = self.root
            new_node.right = self.root.right
            self.root.right = None
        
        self.root = new_node

    def delete(self, key):
        """
        Deletes a key from the tree.

        If the key is not found, the tree is splayed based on the search path,
        but no node is removed.
        """
        if not self.root:
            return

        # Splay the tree to bring the node to delete (if it exists) to the root
        self._splay(key)

        # If the key is not in the tree, do nothing
        if self.root.key != key:
            return

        # At this point, the node to delete is the root
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If there's no left subtree, the right subtree becomes the new tree
            self.root = right_subtree
        else:
            # The left subtree becomes the temporary root
            self.root = left_subtree
            # Splay the temporary tree for the key we are deleting. This will
            # bring the maximum element of the left subtree to its root.
            self._splay(key)
            # The new root (max of left subtree) has no right child, so we
            # can attach the original right subtree there.
            self.root.right = right_subtree

    def search(self, key):
        """
        Searches for a key in the tree.

        This operation splays the found node (or its would-be parent if not
        found) to the root of the tree.
        
        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False
            
        self._splay(key)
        
        # After splaying, the key will be at the root if it exists
        return self.root.key == key

File: generated_code/GeminiPro/Java/086_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-adjusting binary search tree with the property that
 * recently accessed elements are quick to access again. After a search, insertion,
 * or deletion, the accessed node is moved to the root of the tree through a series of
 * rotations in an operation called "splaying".
 */
public class SplayTree {

    /**
     * Inner class to represent a node in the splay tree.
     */
    private class Node {
        int key;
        Node left, right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty SplayTree.
     */
    public SplayTree() {
        this.root = null;
    }

    // Helper function to perform a right rotation on the subtree rooted with y.
    private Node rightRotate(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    // Helper function to perform a left rotation on the subtree rooted with x.
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The main splay operation. It brings the node with the given key
     * (or the last accessed node if the key is not present) to the root of the subtree.
     * This is a recursive, bottom-up splaying implementation.
     *
     * @param node The root of the subtree to splay.
     * @param key  The key to splay around.
     * @return The new root of the splayed subtree.
     */
    private Node splay(Node node, int key) {
        // Base case: node is null or key is present at root
        if (node == null || node.key == key) {
            return node;
        }

        // Key lies in the left subtree
        if (key < node.key) {
            // Key is not in the tree, we are done with the search path
            if (node.left == null) {
                return node;
            }

            // Zig-Zig (Left Left)
            if (key < node.left.key) {
                // Recursively bring the key as root of left-left
                node.left.left = splay(node.left.left, key);
                // Do first rotation for node
                node = rightRotate(node);
            }
            // Zig-Zag (Left Right)
            else if (key > node.left.key) {
                // Recursively bring the key as root of left-right
                node.left.right = splay(node.left.right, key);
                // Do first rotation for node.left
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }

            // Do second rotation for node (for both Zig-Zig and Zig-Zag)
            // or the single rotation for Zig
            return (node.left == null) ? node : rightRotate(node);
        }
        // Key lies in the right subtree
        else {
            // Key is not in the tree, we are done with the search path
            if (node.right == null) {
                return node;
            }

            // Zig-Zag (Right Left)
            if (key < node.right.key) {
                // Recursively bring the key as root of right-left
                node.right.left = splay(node.right.left, key);
                // Do first rotation for node.right
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right Right)
            else if (key > node.right.key) {
                // Recursively bring the key as root of right-right
                node.right.right = splay(node.right.right, key);
                // Do first rotation for node
                node = leftRotate(node);
            }

            // Do second rotation for node (for both Zig-Zig and Zig-Zag)
            // or the single rotation for Zig
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    /**
     * Searches for a key in the tree.
     * This operation performs splaying on the accessed key, moving it (or the last
     * accessed node on the search path) to the root.
     *
     * @param key The key to search for.
     * @return The key as an Integer object if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return root.key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree. Duplicates are not allowed.
     * After insertion, the new node becomes the root of the tree.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree to bring the closest node to the root
        root = splay(root, key);

        // If key is already present, do nothing (it's a set)
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        // If the new key is smaller than the root's key,
        // make the new node the new root. The old root becomes the new root's right child.
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        }
        // If the new key is greater than the root's key,
        // make the new node the new root. The old root becomes the new root's left child.
        else {
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        // If the tree is empty, there is nothing to delete
        if (root == null) {
            return;
        }

        // Splay the tree to bring the node to be deleted (if it exists) to the root
        root = splay(root, key);

        // If the key is not in the tree, do nothing
        if (root.key != key) {
            return;
        }

        // Now, the node to be deleted is the root.
        // We need to merge its left and right subtrees.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree
            root = rightSubtree;
        } else {
            // Splay the left subtree for the key 'key'. This brings the maximum element
            // in the left subtree (the predecessor of 'key') to its root.
            Node newRoot = splay(leftSubtree, key);

            // The new root's right child is guaranteed to be null (as it's the max element).
            // We can now attach the original right subtree there.
            newRoot.right = rightSubtree;
            root = newRoot;
        }
    }
}